rpc/rpcutils/authcodec.go

```diff
@@ -0,0 +1,240 @@
+	headerLenBuf := make([]byte, HEADER_LEN_BYTES)
+	n, err := a.conn.Read(headerLenBuf)
+	if err != nil {
+		return err
+	}
+	if n != HEADER_LEN_BYTES {
+		return ErrReadingHeader
+	}
+
+	headerLength := endian.Uint32(headerLenBuf)
+
+	// Now read the header
+	var header []byte
+	if headerLength > 0 {
+		header = make([]byte, headerLength)
+		n, err = a.conn.Read(header)
+		if err != nil {
+			return err
+		}
+		if uint32(n) != headerLength {
+			return ErrReadingHeader
+		}
+	}
+
+	// Now let the underlying codec read the rest
+	if err := a.wrappedcodec.ReadRequestHeader(r); err != nil {
+		return err
+	}
+
+	// Now we can get the method name from r and authenticate if required
+	if requiresAuthentication(r.ServiceMethod) {
+		ident, err := a.parser.ParseHeader(header)
+		if err != nil {
+			return err
+		}
+
+		if requiresAdmin(r.ServiceMethod) && !ident.HasAdminAccess() {
+			return ErrNoAdmin
+		}
+
+		//TODO: save the identity so we can inject it into the request body later
+
+	}
+	return nil
+}
+
+// Decodes the request and populates the body object with the body of the request
+//  We don't change anything here, just let the underlying codec handle it.
+//  This always gets called after ReadRequestHeader
+func (a *AuthServerCodec) ReadRequestBody(body interface{}) error {
+	// TODO: Use reflection and add the identity to the body if necessary
+	return a.wrappedcodec.ReadRequestBody(body)
+}
+
+//  Encodes the response before sending it back down to the client.
+//  We don't change anything here, just let the underlying codec handle it.
+func (a *AuthServerCodec) WriteResponse(r *rpc.Response, body interface{}) error {
+	return a.wrappedcodec.WriteResponse(r, body)
+}
+
+// Closes the connection on the server side
+//  We don't change anything here, just let the underlying codec handle it.
+func (a *AuthServerCodec) Close() error {
+	return a.wrappedcodec.Close()
+}
+
+// Client Codec
+type AuthClientCodec struct {
+	conn          io.ReadWriteCloser
+	wrappedcodec  rpc.ClientCodec
+	headerBuilder auth.RPCHeaderBuilder
+	mutex         sync.Mutex
+}
+
+func NewDefaultAuthClientCodec(conn io.ReadWriteCloser) rpc.ClientCodec {
+	return NewAuthClientCodec(conn, jsonrpc.NewClientCodec(conn), &auth.RPCHeaderHandler{})
+}
+
+func NewAuthClientCodec(conn io.ReadWriteCloser, codecToWrap rpc.ClientCodec, headerBuilder auth.RPCHeaderBuilder) rpc.ClientCodec {
+	return &AuthClientCodec{
+		conn:          conn,
+		wrappedcodec:  codecToWrap,
+		headerBuilder: headerBuilder,
+	}
+}
+
+// Encodes the request and sends it to the server.
+// This implementation gets an auth header when appropriate, and writes it to the stream
+//  before letting the underlying codec send the rest of the request.
+func (a *AuthClientCodec) WriteRequest(r *rpc.Request, body interface{}) error {
+	var (
+		header []byte = []byte{}
+		err    error
+	)
+
+	if requiresAuthentication(r.ServiceMethod) {
+		header, err = a.headerBuilder.BuildHeader()
```

Comment:
Yo, you gots to pass the request in there and sign it, or your header is replayable